### ISSUE TITLE
Printing lor

### ISSUE THREAD

**Issue by ruiminhe** [coq/coq#12983]
Hi,

In 8.12, from a fresh screen,

Require Import Numbers.Cyclic.Int63.Int63. Locate Int63.lor. works but 
Require Import Numbers.Cyclic.Int63.Int63. Locate lor. yields

Syntax error: 'Ltac' or [locatable] expected after 'Locate' (in [vernac:command]).

It seems like this is an issue because lor is simultaneously a term as well as a notation, i.e. from

Primitive lor := #int63_lor.
Infix "lor" := lor (at level 40, left associativity) : int63_scope.

**Comment by herbelin** [comment]
Indeed, module `Int63` modifies the parser to make `lor` an infix operator and this breaks the use of `lor` as a regular identifier.

I believe that this is intended and not a bug. Cc @maximedenes, @proux01, @JasonGross.

**Comment by proux01** [comment]
This seems intended to me too.

However
```Coq
Locate "lor".
```
yields
```
Notation
"x 'lor' y" := lor x y : int63_scope
```
maybe one would find an answer such as
```
Notation
"x 'lor' y" := Int63.lor x y : int63_scope
```
more convenient.

**Comment by herbelin** [comment]
> ```
> Notation
> "x 'lor' y" := Int63.lor x y : int63_scope
> ```

Indeed! Then `shortest_qualid` should avoid the identifiers which are registered as tokens and satisfy `CLexer.is_keyword`.

**Comment by proux01** [comment]
That would be great

**Comment by ruiminhe** [comment]
Got it. Sounds like there will be no way of locating the term instead of the notation then, because

Locate Term lor

doesn't work either.

**Comment by herbelin** [comment]
> That would be great

We would then need help from @ejgallego: `parsing` comes after `library`, so `is_keyword` is not available in `short_qualid`. Shall we use a hook? Move the table of keywords in `library`? Add an extra table in, say, `Nameops` or `Nametab`, to keep a copy of those idents that are keywords?

I believe I'd be leaning by default to proposing option 3.

**Comment by ejgallego** [comment]
I dunno; morally it seems like `Nametab` should live in `interp`, as lower layers do have little concern with the interpretation, and indeed printing does want to access the notation machinery in `extern` etc....

I guess a table is a bad idea too as it will duplicate info, maybe `shortest_qualid` could take an extra parameter ?

**Comment by herbelin** [comment]
> morally it seems like Nametab should live in interp

Good point.

> maybe shortest_qualid could take an extra parameter ?

Maybe a bit intrusive change, regarding the importance of the feature, but why not.

**Comment by ejgallego** [comment]
> Maybe a bit intrusive change, regarding the importance of the feature, but why not.

I dunno, we could introduce a new function if compat is your worry; other approaches [hook or table] would be even more intrusive I think.

**Comment by herbelin** [comment]
I meant intrusive for the end user of the API if all calls to `short_qualid` have to be modified. But, rethinking at your words, I think I misunderstood what kind of extra parameters you want to pass, because there are calls to `short_qualid` already in `interp`, and `interp` does not know `parsing`.

**Comment by ejgallego** [comment]
Doesn't interp have enough information already as avoid having to look into the low-level `is_keyword` table?

Actually, is using this table the right thing to do here? I'd guess that keywords are a more general concept than a notation, right?

**Comment by ejgallego** [comment]
I think so, IMHO here we are talking about basically how externalization is going to undo identifier resolution for printing so it seems natural that it could consult notation data; I guess there would be no problem having externalization access the parser if that would be useful, but it seems to me that `is_keyword` may not be the best match; but indeed that's subtle.
